pkgs/build-support/rust/rust-utils.nix

```diff
@@ -0,0 +1,373 @@
+
+         mkdir -p target/build/${crateName}.out
+         export RUST_BACKTRACE=1
+         BUILD_OUT_DIR="-L $OUT_DIR"
+         mkdir -p $OUT_DIR
+         target/build/${crateName}/build_script_build > target/build/${crateName}.opt
+         set +e
+         EXTRA_BUILD=$(sed -n "s/^cargo:rustc-flags=\(.*\)/\1/p" target/build/${crateName}.opt | tr '\n' ' ')
+         EXTRA_FEATURES=$(sed -n "s/^cargo:rustc-cfg=\(.*\)/--cfg \1/p" target/build/${crateName}.opt | tr '\n' ' ')
+         EXTRA_LINK=$(sed -n "s/^cargo:rustc-link-lib=\(.*\)/\1/p" target/build/${crateName}.opt | tr '\n' ' ')
+         EXTRA_LINK_SEARCH=$(sed -n "s/^cargo:rustc-link-search=\(.*\)/\1/p" target/build/${crateName}.opt | tr '\n' ' ')
+         CRATENAME=$(echo ${crateName} | sed -e "s/\(.*\)-sys$/\U\1/")
+         grep -P "^cargo:(?!(rustc-|warning=|rerun-if-changed=|rerun-if-env-changed))" target/build/${crateName}.opt \
+           | sed -e "s/cargo:\([^=]*\)=\(.*\)/export DEP_$(echo $CRATENAME)_\U\1\E=\2/" > target/env
+
+         set -e
+         if [ -n "$(ls target/build/${crateName}.out)" ]; then
+
+            if [ -e "${libPath}" ] ; then
+               cp -r target/build/${crateName}.out/* $(dirname ${libPath}) #*/
+            else
+               cp -r target/build/${crateName}.out/* src #*/
+            fi
+         fi
+      fi
+
+      EXTRA_LIB=""
+      CRATE_NAME=$(echo ${libName} | sed -e "s/-/_/g")
+
+      if [ -e target/link_ ]; then
+        EXTRA_BUILD="$(cat target/link_) $EXTRA_BUILD"
+      fi
+
+      if [ -e "${libPath}" ] ; then
+         build_lib ${libPath}
+      elif [ -e src/lib.rs ] ; then
+         build_lib src/lib.rs
+      elif [ -e src/${libName}.rs ] ; then
+         build_lib src/${libName}.rs
+      fi
+
+      echo "$EXTRA_LINK_SEARCH" | while read i; do
+         if [ ! -z "$i" ]; then
+           for lib in $i; do
+             echo "-L $lib" >> target/link
+             L=$(echo $lib | sed -e "s#$(pwd)/target/build#$out/lib#")
+             echo "-L $L" >> target/link.final
+           done
+         fi
+      done
+      echo "$EXTRA_LINK" | while read i; do
+         if [ ! -z "$i" ]; then
+           for lib in $i; do
+             echo "-l $lib" >> target/link
+             echo "-l $lib" >> target/link.final
+           done
+         fi
+      done
+
+      if [ -e target/link ]; then
+         sort -u target/link.final > target/link.final.sorted
+         mv target/link.final.sorted target/link.final
+         sort -u target/link > target/link.sorted
+         mv target/link.sorted target/link
+
+         tr '\n' ' ' < target/link > target/link_
+         LINK=$(cat target/link_)
+       fi
+
+      mkdir -p target/bin
+      echo "${crateBin}" | sed -n 1'p' | tr ',' '\n' | while read BIN; do
+         if [ ! -z "$BIN" ]; then
+           build_bin $BIN
+         fi
+      done
+      if [[ (-z "${crateBin}") && (-e src/main.rs) ]]; then
+         build_bin ${crateName} src/main.rs
+      fi
```

Comment:
Cargo will also build any file in `bin/*.rs` as executable. The following code will enable that:

```diff
diff --git a/pkgs/build-support/rust/rust-utils.nix b/pkgs/build-support/rust/rust-utils.nix
index 0e9458bab11..5070186f07c 100644
--- a/pkgs/build-support/rust/rust-utils.nix
+++ b/pkgs/build-support/rust/rust-utils.nix
@@ -257,9 +257,14 @@ let buildCrate = { crateName, crateVersion, crateAuthors, buildDependencies,
            build_bin $BIN
          fi
       done
-      if [[ (-z "${crateBin}") && (-e src/main.rs) ]]; then
-         build_bin ${crateName} src/main.rs
-      fi
+      ${lib.optionalString (crateBin == "") ''
+        if [[ -e src/main.rs ]]; then
+          build_bin ${crateName} src/main.rs
+        fi
+        for i in src/bin/*.rs; do
+          build_bin "$(basename $i .rs)" "$i"
+        done
+      ''}
       # Remove object files to avoid "wrong ELF type"
       find target -type f -name "*.o" -print0 | xargs -0 rm -f
       runHook postBuild
```

tested with [ofborg](https://github.com/grahamc/ofborg)

Reply:
Cargo doesn't always build `bin/*.rs`.  https://github.com/rust-lang/cargo/issues/4013

Reply:
This is a nightmare. How do we handle this? Without good multiple-binary support, this patchset feels not useful.

Reply:
I mentioned the FYI only because I don't fully understand how this patch set works... yet, and wanted to make sure this wasn't overlooked.

It may already be handled.  I have a Rust project I'm packaging to run on NixOS that has two `[[bin]]` sections in the Cargo.toml.  After running my Cargo.lock through carnix, my nix package builds both binaries in the crate just fine.

The best I can figure is carnix emulates cargo's behavior here.  Carnix does parse the Cargo.toml and if it finds `[[bin]]` sections, it provides a list of bin files as `crateBin`.  Otherwise, it provides and empty list that will trigger Mic92's for loop.  There's a bit of speculation on my part in there.

Reply:
Oh I see, if I declared my binaries in the Cargo.toml it wouldn't require this patch? Great! Makes sense, thank you for the clarification.

Reply:
To better illustrate my example, here's a gist \[[1](https://gist.github.com/boxofrox/59b0a7fe4cce97e9c5753aee50bd25d9)]  of the carnix output for my project.  Towards the bottom is a more condensed version with a layout of my project directory.  The `cargoBin` looks as I'd expect.

I'll run a few more tests and report back.

\[1]: https://gist.github.com/boxofrox/59b0a7fe4cce97e9c5753aee50bd25d9

Reply:
@boxofrox yes, you are right, we do the same as cargo and skip building `main.rs` and `bin/*.rs` in the presents of `[[bin]]` section.

Reply:
Great.  And the tests I ran--disabling some and all `[[bin]]` sections--produced nix files that reflect this behavior when ran through `nix-build`.  This patch completes the imitation of cargo.  Cheers.